// cozydb/lib/index.js
// Generated by CoffeeScript 1.8.0
(function() {
  var Controller, CozyModel, Model, NoSchema, api, defaultRequests, emit, log,
    __hasProp = {}.hasOwnProperty,
    __extends = function(child, parent) { for (var key in parent) { if (__hasProp.call(parent, key)) child[key] = parent[key]; } function ctor() { this.constructor = child; } ctor.prototype = parent.prototype; child.prototype = new ctor(); child.__super__ = parent.prototype; return child; };

  module.exports.Model = Model = require('./model');

  module.exports.CozyModel = CozyModel = require('./cozymodel');

  module.exports.SimpleController = Controller = require('./controller');

  NoSchema = require('./utils/type_checking').NoSchema;

  module.exports.NoSchema = NoSchema;

  emit = function() {};

  module.exports.defaultRequests = defaultRequests = {
    all: function(doc) {
      return emit(doc._id, doc);
    },
    tags: function(doc) {
      var tag, _i, _len, _ref, _results;
      _ref = doc.tags || [];
      _results = [];
      for (_i = 0, _len = _ref.length; _i < _len; _i++) {
        tag = _ref[_i];
        _results.push(emit(tag, doc));
      }
      return _results;
    },
    by: function(field) {
      return (function(doc) {
        return emit(doc.FIELD, doc);
      }).toString().replace('FIELD', field);
    }
  };

  module.exports.api = api = require('./api');

  module.exports.getModel = function(name, schema) {
    var ClassFromGetModel, klass;
    klass = ClassFromGetModel = (function(_super) {
      __extends(ClassFromGetModel, _super);

      function ClassFromGetModel() {
        return ClassFromGetModel.__super__.constructor.apply(this, arguments);
      }

      ClassFromGetModel.schema = schema;

      return ClassFromGetModel;

    })(CozyModel);
    klass.displayName = klass.name = name;
    klass.toString = function() {
      return "" + name + "Constructor";
    };
    klass.docType = name;
    return klass;
  };

  log = function() {
    if (process.env.NODE_ENV === 'test') {
      return null;
    }
    return console.log.apply(console, arguments);
  };

  module.exports.configure = function(options, app, callback) {
    var Pouch, PouchModel, docType, err, model, modelPath, requestDefinition, requestDefinitions, requestName, requests, requestsToSave, step;
    if (callback == null) {
      callback = function() {};
    }
    if (typeof options === 'string') {
      options = {
        root: options
      };
    }
    if (options.db || options.dbName) {
      try {
        Pouch = require('pouchdb');
        PouchModel = require('./pouchmodel');
        module.exports.CozyModel = CozyModel = PouchModel;
        if (options.db) {
          PouchModel.db = options.db;
        } else {
          if (options.dbName == null) {
            options.dbName = process.env.POUCHDB_NAME || 'cozy';
          }
          PouchModel.db = new Pouch(options.dbName);
        }
      } catch (_error) {
        err = _error;
        console.log(err);
        return callback(err);
      }
    }
    modelPath = "" + options.root + "/server/models/";
    try {
      requests = require(modelPath + "requests");
    } catch (_error) {
      err = _error;
      log("failed to load requests file", err);
      return callback(err);
    }
    requestsToSave = [];
    for (docType in requests) {
      requestDefinitions = requests[docType];
      model = require(modelPath + docType);
      for (requestName in requestDefinitions) {
        requestDefinition = requestDefinitions[requestName];
        requestsToSave.push({
          model: model,
          requestName: requestName,
          requestDefinition: requestDefinition
        });
      }
    }
    requestsToSave.push({
      model: api.CozyInstance,
      requestName: 'all',
      requestDefinition: defaultRequests.all
    });
    requestsToSave.push({
      model: api.CozyUser,
      requestName: 'all',
      requestDefinition: defaultRequests.all
    });
    step = function(i) {
      var _ref;
      if (i == null) {
        i = 0;
      }
      _ref = requestsToSave[i], model = _ref.model, requestName = _ref.requestName, requestDefinition = _ref.requestDefinition;
      log("" + (model.getDocType()) + " - " + requestName + " request creation...");
      return model.defineRequest(requestName, requestDefinition, function(err) {
        if (err) {
          log("failed", err);
          log("A request creation failed, abandon.");
          return callback(err);
        } else if (i + 1 >= requestsToSave.length) {
          log("requests creation complete");
          return callback(null);
        } else {
          log("succeeded");
          return step(i + 1);
        }
      });
    };
    return step(0);
  };

}).call(this);
